lib/cloud_controller/deployment_updater/actions/finalize.rb

```diff
@@ -0,0 +1,55 @@
+require 'cloud_controller/deployment_updater/actions/cleanup_web_processes'
+module VCAP::CloudController
+  module DeploymentUpdater
+    module Actions
+      class Finalize
+        attr_reader :deployment, :deploying_web_process, :app
+
+        def initialize(deployment)
+          @deployment = deployment
+          @app = deployment.app
+          @deploying_web_process = deployment.deploying_web_process
+        end
+
+        def call
+          promote_deploying_web_process
+
+          CleanupWebProcesses.new(deployment, deploying_web_process).call
+
+          update_non_web_processes
+          restart_non_web_processes
+          deployment.update(
+            state: DeploymentModel::DEPLOYED_STATE,
+            status_value: DeploymentModel::FINALIZED_STATUS_VALUE,
+            status_reason: DeploymentModel::DEPLOYED_STATUS_REASON
+          )
+        end
+
+        private
+
+        def promote_deploying_web_process
+          deploying_web_process.update(type: ProcessTypes::WEB)
+        end
+
+        def update_non_web_processes
+          return if deploying_web_process.revision.nil?
```

Comment:
**question** Do we know what happens here if revisions are disable for the app?

Reply:
I think so.

Reply:
non-blocking, but did we ever understand this?

Reply:
So this method doesn't have anything to do with the droplet, just gets the command from the revision. If revisions are disabled, then the processes get their commands from the Procfile or app manifest, probably in the apply manifest job (not sure about when the Procfile gets read).